update document for FlavorFungibility (#1226)
kubernetes-sigs/kueue: site/content/en/docs/concepts/cluster_queue.md

- Workloads with the lowest priority.
- Workloads that have been admitted more recently.

> ## FlavorFungibility

Reviewer: Add a note in the `Flavors and borrowing semantics` that some semantics can be modified based on the `flavorFungibility` field, and add a link to this section.

When there is not enough nominal quota of resources in a ResourceFlavor, the incoming workload can borrow or preempt in ClusterQueue or Cohort. 

> By default the incoming workload will stop trying the next flavor if it can get enough resource by borrowing. 

Reviewer 2: I think we should clarify what is `the next flavor`.

Reviewer: Yeah, we can remind readers that kueue evaluates the flavors in a ClusterQueue in order.

Reviewer: But generally, we introduce the possible options first, followed by what's the default.

Reviewer 2: agree.

Author: Sure, I will update this

> By default the incoming workload will stop trying the next flavor if it can get enough resource by borrowing. 

Reviewer 3: ```suggestion
By default, the incoming workload  stops trying the next flavor if the workload can get enough borrowed resources. 
```

I'm making a lot of tech assumptions. Feel free to disregard if the meaning was affected

> And preemption will be triggered after all ResourceFlavors were considered.

Reviewer: ```suggestion
And preemption will be triggered only after Kueue determines that the remaining ResourceFlavors can't fit the workload.
```

> And preemption will be triggered after all ResourceFlavors were considered.

Reviewer 3: ```suggestion
And Kueue triggers preemption after considering all ResourceFlavors.
```
Changing to active voice. I assume the subject of the sentence is Kueue, but feel free to change it,

> ClusterQueues can also try preemption whenever preemption might help before trying next ResourceFlavor or only borrowing after all ResourceFlavors have been considered by setting `FlavorFungibility`.

Reviewer: ```suggestion
You can configure ClusterQueues to try preemption whenever preemption might help before trying next ResourceFlavor or only borrowing after all ResourceFlavors have been considered by setting `FlavorFungibility`.
```

A configuration for a ClusterQueue that changes this behavior looks like the following:

> ```

Reviewer 2: ```suggestion

```yaml
```

apiVersion: kueue.x-k8s.io/v1beta1
kind: ClusterQueue
metadata:
  name: "team-a-cq"
spec:
  flavorFungibility:
    whenCanBorrow: TryNextFlavor
    whenCanPreempt: Preempt
```

> The fields above have the following semantics:

Reviewer 3: ```suggestion
The fields above do the following:
```

Author: https://github.com/kubernetes-sigs/kueue/blob/40bd411beddb5bd415d880a0c841c92ababc0458/site/content/en/docs/concepts/cluster_queue.md?plain=1#L324C47-L324C47

I have also made modifications to the identical expression within the text.

- `whenCanBorrow` determines whether a workload should stop finding a better assignment if it can get enough resource by borrowing in current ResourceFlavor. The possible values are:

>   - `Borrow` (default): ClusterQueue stops to find a better assignment.

Reviewer 2: ```suggestion
  - `Borrow` (default): ClusterQueue stops finding a better assignment.
```

>   - `TryNextFlavor`: ClusterQueue will try the next ResourceFlavor to see if the workload can get a better assignment.

Reviewer 3: ```suggestion
  - `TryNextFlavor`: ClusterQueue tries the next ResourceFlavor to see if the workload can get a better assignment.
```

- `whenCanPreempt` determines whether a workload should try preemtion in current ResourceFlavor before try the next one. The possible values are:

>   - `Preempt`: ClusterQueue stops to try preempting in current ResourceFlavor and start from the next one if preempting failed.

Reviewer 2: ```suggestion
  - `Preempt`: ClusterQueue stops trying preemption in current ResourceFlavor and starts from the next one if preempting failed.
```

>   - `TryNextFlavor` (default): ClusterQueue will try the next ResourceFlavor to see if the workload can fit in the ResourceFlavor.

Reviewer 3: ```suggestion
  - `TryNextFlavor` (default): ClusterQueue tries the next ResourceFlavor to see if the workload can fit in the ResourceFlavor.
```

> Noted that if `whenCanBorrow` is `Borrow` and `whenCanPreempt` is `Preempt`, or both `whenCanBorrow` and `whenCanPreempt` is `TryNextFlavor`, borrowing has a higher priority than preemption.

Reviewer: This is highly confusing. Maybe simplify by saying that Kueue tries to avoid preemptions by borrowing, whenever possible and the configured policy allows it?

> Noted that if `whenCanBorrow` is `Borrow` and `whenCanPreempt` is `Preempt`, or both `whenCanBorrow` and `whenCanPreempt` is `TryNextFlavor`, borrowing has a higher priority than preemption.

Reviewer 3: ```suggestion
To instruct Kueue to attempt to borrow first instead of preempt, you can use any of the following value combinations:
*  `whenCanBorrow = "Borrow"` and `whenCanPreempt = "Preempt"`
*  `whenCanBorrow = "TryNextFlavor"` and `whenCanPreempt = "TryNextFlavor"`
```

Author: I think that such expression is sufficient to convey the meaning: 

```
Note that whenever possible and the configured policy allows, Kueue tries to avoid preemptions by borrowing.
```

- Workloads with the lowest priority.
- Workloads that have been admitted more recently.

## FlavorFungibility

When there is not enough nominal quota of resources in a ResourceFlavor, the incoming workload can borrow or preempt in ClusterQueue or Cohort.

Kueue evaluates the flavors in a ClusterQueue in order. You can configure ClusterQueues to try preemption whenever preemption might help before trying next ResourceFlavor or only borrowing after all ResourceFlavors have been considered by setting `FlavorFungibility`.

> A configuration for a ClusterQueue that changes this behavior looks like the following:

Reviewer: ```suggestion
A configuration for a ClusterQueue that configures this behavior looks like the following:
```

kind: ClusterQueue
metadata:
  name: "team-a-cq"
spec:
  flavorFungibility:
    whenCanBorrow: TryNextFlavor
    whenCanPreempt: Preempt
```

The fields above do the following:
- `whenCanBorrow` determines whether a workload should stop finding a better assignment if it can get enough resource by borrowing in current ResourceFlavor. The possible values are:
  - `Borrow` (default): ClusterQueue stops finding a better assignment.
  - `TryNextFlavor`: ClusterQueue tries the next ResourceFlavor to see if the workload can get a better assignment.
- `whenCanPreempt` determines whether a workload should try preemtion in current ResourceFlavor before try the next one. The possible values are:
  - `Preempt`: ClusterQueue stops trying preemption in current ResourceFlavor and starts from the next one if preempting failed.
  - `TryNextFlavor` (default): ClusterQueue tries the next ResourceFlavor to see if the workload can fit in the ResourceFlavor.

By default, the incoming workload stops trying the next flavor if the workload can get enough borrowed resources. 
And Kueue triggers preemption only after Kueue determines that the remaining ResourceFlavors can't fit the workload.

> Note that whenever possible and the configured policy allows, Kueue tries to avoid preemptions by borrowing. 

Reviewer: ```suggestion
Note that, whenever possible and when the configured policy allows it, Kueue avoids preemptions if it can fit a Workload by borrowing. 
```

- Workloads with the lowest priority.
- Workloads that have been admitted more recently.

## FlavorFungibility

When there is not enough nominal quota of resources in a ResourceFlavor, the incoming Workload can borrow
quota or preempt running Workloads in the ClusterQueue or Cohort.

Kueue evaluates the flavors in a ClusterQueue in order. You can influence whether to prioritize
preemptions or borrowing in a flavor before trying to accommodate the Workload in the next flavor, by
setting the `flavorFungibility` field.

> A configuration for a ClusterQueue that configures this behavior looks like the following:

Reviewer 4: nit:

```suggestion
The definition of a ClusterQueue that configures this behavior looks like the following:
```
"configuration  .. that configures" sounds a bit strange to me.
